### DISCUSSION TITLE
How to compute the contact Jacobian?

### DISCUSSION THREAD

**Discussion by xiaosuyang** [stack-of-tasks/pinocchio#2201] [General]
Hello pinocchio community. I'm working on Biped robot 's walking control.
In the formula:
![image](https://github.com/stack-of-tasks/pinocchio/assets/55594017/f662c4a2-343a-4a9b-9594-229cbd851cd5)

How to calculate the contac jacobian using pinocchio?
Can I Establish a coordinate system under the foot surface and then use `getFrameJacobian(model,data, footframe_index, LOCAL)` or use the `getFrameJacobian(model,data,footframe_index,LOCAL_WORLD_ALIGNED)` ?

**Comment by ManifoldFR** [comment]
Hello mate,

You are in the right headspace - one would indeed usually go about this by creating a `pin.Frame` attached to the foot (adding it to the `model` object) and grabbing the Jacobian for that frame.

However, `pin.getFrameJacobian()` actually requires that `pin.computeJointJacobians()` and `pin.updateFramePlacements()` were called first - because `getFrameJacobian()` composes the parent joint's jacobian with an appropriate matrix to account for the frame's relative placement.

As a shortcut (which is more efficient _for a single Jacobian_ because it will skip superfluous computations) you should use `pin.computeFrameJacobian()` instead. Have fun !


For good measure - you didn't have to close #2200 (which I've deleted) and open this one, editing the previous one was enough (we get a bunch of e-mail notifications when someone opens a discussion/issue 😅)

**Reply by xiaosuyang** [reply]
Thank you for your guidance.  But in above formular, if the external force $f_r$ is expressed in world frame,  which frame should we use to map the $f_r$ to the joint space, LOCAL_WORLD_ALIGNED or the WORLD frame?

**Reply by xiaosuyang** [reply]
I use the MPC to compute the  reaction force($f_r\in F^6$) of a point of the stance foot. And I compute the geometric jacobian(Use the biped 's torsor frame as a fixed base) $J (6\times6)$ 6 dof a leg to map the $f_r$ to the joint, and it seems that it work. But the geometric jacobian $J$ is not the contact jacobian including the floating base joint， That was confusing.

**Comment by cmastalli** [comment]
Hi @xiaosuyang,

If you are interested in computing the "dynamics", then the most efficient option is to map your contact forces into its body frame. ABA and RNEA algorithms recursively account for such a spatial force without involving an inefficient matrix-vector multiplication.

Computing contact Jacobians is needed when you need to impose "contact constraints".
